app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,229 @@
+package org.kiwix.kiwixmobile.zim_manager.local_file_transfer;
+
+import android.content.BroadcastReceiver;
+import android.content.Context;
+import android.content.IntentFilter;
+import android.net.wifi.WpsInfo;
+import android.net.wifi.p2p.WifiP2pConfig;
+import android.net.wifi.p2p.WifiP2pDevice;
+import android.net.wifi.p2p.WifiP2pDeviceList;
+import android.net.wifi.p2p.WifiP2pInfo;
+import android.net.wifi.p2p.WifiP2pManager;
+import android.os.Build;
+import android.util.Log;
+import android.widget.Toast;
+import androidx.annotation.NonNull;
+import java.net.InetAddress;
+import org.kiwix.kiwixmobile.R;
+
+import static android.os.Looper.getMainLooper;
+import static org.kiwix.kiwixmobile.zim_manager.local_file_transfer.LocalFileTransferActivity.showToast;
+
+
+/**
+ * Manager for the Wifi-P2p API, used in the local file transfer module
+ * */
+public class WifiDirectManager implements WifiP2pManager.ChannelListener, WifiP2pManager.PeerListListener, WifiP2pManager.ConnectionInfoListener {
+
+  private static final String TAG = "WifiDirectManager";
+
+  @NonNull LocalFileTransferActivity activity;
+
+  /* Variables related to the WiFi P2P API */
+  private boolean wifiP2pEnabled = false; // Whether WiFi has been enabled or not
+  private boolean retryChannel = false;   // Whether channel has retried connecting previously
+
+  private WifiP2pManager manager;         // Overall manager of Wifi p2p connections for the module
+  private WifiP2pManager.Channel channel;
+  // Connects the module to device's underlying Wifi p2p framework
+
+  private final IntentFilter intentFilter = new IntentFilter();
+  // For specifying broadcasts (of the P2P API) that the module needs to respond to
+  private BroadcastReceiver receiver = null; // For receiving the broadcasts given by above filter
+
+  private WifiP2pDevice userDevice;   // Represents the device on which the app is running
+  private WifiP2pInfo groupInfo;      // Corresponds to P2P group formed between the two devices
+
+  private WifiP2pDevice senderSelectedPeerDevice = null;
+
+  public WifiDirectManager(@NonNull LocalFileTransferActivity activity) {
```

Comment:
You should not have the concrete class LocalFileTransferActivity, you can just inject Activity in the constructor. This will force you to communicate via callbacks instead of directly invoking LFTA methods.

 You can also inject WifiP2pManager  but you will have to supply the binding, see `ApplicationModule`

Reply:
> You should not have the concrete class LocalFileTransferActivity, you can just inject Activity in the constructor

Like [this](https://github.com/kiwix/kiwix-android/pull/1276/commits/da141c862280b4362e80042adf54139c0b8415a8#diff-7e01f6f425daaca67175b74c839c6af2R84) right?

> You can also inject WifiP2pManager but you will have to supply the binding, see ApplicationModule

After the review changes, I think the current injectable WDM will meet the needs for file transfer? Or should I create the binding for WifiP2pManager all the same?

Reply:
Inject WifiP2PManager just like you did the Location.... thingy I forgot the name of.
The more and better you organise your constructor parameters you start to see them describing what the class actually does "WifiDirectManager needs a P2PManager"